warehouse/internal/repo/upload.go

```diff
@@ -680,3 +687,362 @@ func (uploads *Uploads) LastCreatedAt(ctx context.Context, sourceID, destination
 
 	return createdAt.Time, nil
 }
+
+func (uploads *Uploads) SyncsInfoForMultiTenant(ctx context.Context, limit, offset int, opts model.SyncUploadOptions) ([]model.UploadInfo, int64, error) {
+	syncUploadInfos, totalUploads, err := uploads.syncsInfo(ctx, limit, offset, opts, true)
+	if err != nil {
+		return nil, 0, fmt.Errorf("syncs upload info for multi tenant: %w", err)
+	}
+	if len(syncUploadInfos) != 0 {
+		return syncUploadInfos, totalUploads, nil
+	}
+
+	// Getting the syncs count in case, we were not able to get the count
+	totalUploads, err = uploads.syncsCount(ctx, opts)
+	if err != nil {
+		return nil, 0, fmt.Errorf("syncs upload count: %w", err)
+	}
+
+	return syncUploadInfos, totalUploads, nil
+}
+
+func (uploads *Uploads) SyncsInfoForNonMultiTenant(ctx context.Context, limit, offset int, opts model.SyncUploadOptions) ([]model.UploadInfo, int64, error) {
+	syncUploadInfos, _, err := uploads.syncsInfo(ctx, limit, offset, opts, false)
+	if err != nil {
+		return nil, 0, fmt.Errorf("syncs upload info for multi tenant: %w", err)
+	}
+
+	totalUploads, err := uploads.syncsCount(ctx, opts)
+	if err != nil {
+		return nil, 0, fmt.Errorf("syncs upload count: %w", err)
+	}
+
+	return syncUploadInfos, totalUploads, nil
+}
+
+func (uploads *Uploads) syncsInfo(ctx context.Context, limit, offset int, opts model.SyncUploadOptions, countOver bool) ([]model.UploadInfo, int64, error) {
+	filterQuery, filterArgs := syncUploadQueryArgs(&opts)
+
+	countOverClause := "0 AS total_uploads"
+	if countOver {
+		countOverClause = "COUNT(*) OVER() AS total_uploads"
+	}
```

Comment:
[question] This part of logic I don't understand, why we return `total_uploads: COUNT(*) OVER()` for multitenant and zero for non-mulitenant ?

Reply:
For non-multi-tenant, we are running a count query instead of embedding it in the same query.
https://github.com/rudderlabs/rudder-server/pull/3814/files#diff-dc7a971c99be83650f101872e58fda0aac0b14297cc4d6376485ed2fec63c8e9R715-R718

Reply:
Dhawal made this improvement when the queries to hostedmted for count(*) were taking time. So he embedded `COUNT(*) OVER() AS total_uploads` in the same query to get the total count.

Reply:
I can probably benchmark the query after this PR and can have a single query for both. wdyt?